### PR TITLE
Fix half precision base support on KEPLER35

### PR DESCRIPTION
**Pull request by e10harvey**

### CODE REVIEW

**Comment by dalg24** [comment]
Add to whitelist

File: core/src/Kokkos_Half.hpp [hunk 1]
```diff
@@ -116,5 +113,7 @@ cast_from_half(half_t val) {
 }  // namespace Experimental
 }  // namespace Kokkos
 
-#endif
-#endif
+#else
+#define KOKKOS_HALF_T_IS_FLOAT false
```
Comment (dalg24) [inline core/src/Kokkos_Half.hpp:117]: This might trip those who do `#ifdef KOKKOS_HALF_T_IS_FLOAT ` out of habit instead of `#if KOKKOS_HALF_T_IS_FLOAT`

Comment (e10harvey) [reply core/src/Kokkos_Half.hpp:117]: I'll defer to @crtrott on this. I can add a `KOKKOS_HALF_T_IS_NOT_FLOAT` define if that's preferred.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 2]
```diff
@@ -47,12 +47,13 @@
 
 #include <Kokkos_Macros.hpp>
 #ifdef KOKKOS_ENABLE_CUDA
+#if !(defined(KOKKOS_COMPILER_CLANG) && KOKKOS_COMPILER_CLANG < 900) && \
+    !defined(KOKKOS_ARCH_KEPLER35)
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:51]: we probably need others here too. My guess is everything below Pascal. That means Kepler37, Maxwell50 and Maxwell52

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:51]: Just try the compilation with all the archs and see which one works

File: core/src/Kokkos_Half.hpp [hunk 3]
```diff
@@ -61,6 +58,7 @@
 // define a fallback implementation here using float
 #ifndef KOKKOS_IMPL_HALF_TYPE_DEFINED
 #define KOKKOS_IMPL_HALF_TYPE_DEFINED
+#define KOKKOS_HALF_T_IS_FLOAT true
```
Comment (crtrott) [inline core/src/Kokkos_Half.hpp:61]: We will discuss this tomorrow.

File: core/src/Kokkos_Half.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/Kokkos_Half.hpp:117]: so we either need the Macro as defined/undefined + the constexpr bool, or just the macro true or false.

**Review by crtrott — CHANGES_REQUESTED** [review]
We definitive need to add more archs here. And we need to discuss the Macro being a bool tomorrow.

**Comment by dalg24** [comment]
Related to #3549

File: core/src/Kokkos_Half.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/Kokkos_Half.hpp:117]: It was brought up at the developer meeting that misspelling the macro in an `#ifdef` is also a risk

Comment (dalg24) [reply core/src/Kokkos_Half.hpp:117]: We decided this is fine for now

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 2 above)
Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:51]: Maxwell52, Maxwell50, Keppler37, Keppler32, Keppler30 fail. Issuing PR.

**Comment by janciesko** [comment]
Superseded by https://github.com/kokkos/kokkos/pull/3571

**Pull request merged**